message/message.go

```diff
@@ -0,0 +1,195 @@
+
+// SignatureVerifier validates Signer signatures
+type SignatureVerifier interface {
+	// Verify checks if the signature of the message is valid
+	Verify(signer, digest, signature []byte) error
+}
+
+// Keccak hash engine for arbitrary input
+type Keccak interface {
+	// Hash returns the Keccak encoding of given data
+	Hash(data []byte) []byte
+}
+
+// Transport is used to gossip consensus messages to the network
+type Transport interface {
+	// MulticastProposal gossips MsgProposal to other consensus peers
+	MulticastProposal(msg *MsgProposal)
+
+	// MulticastPrepare gossips MsgPrepare to other consensus peers
+	MulticastPrepare(msg *MsgPrepare)
+
+	// MulticastCommit gossips MsgCommit to other consensus peers
+	MulticastCommit(msg *MsgCommit)
+
+	// MulticastRoundChange gossips MsgRoundChange to other consensus peers
+	MulticastRoundChange(msg *MsgRoundChange)
+}
+
+// Feed provides Sequencer an asynchronous way to receive consensus messages. In addition to
+// listening for any type of message in any particular view, the higherRounds flag provides an option
+// to include messages from higher rounds as well.
+//
+// CONTRACT:
+//
+// 1. any message is valid:
+//   - no required fields missing (Sender, Signature, getView)
+//   - signature is valid [ recover(ibftMsg.Payload, Signature) == Sender ]
+//
+// 2. all messages are considered unique (there cannot be 2 or more messages from the same sender)
+type Feed interface {
+	// SubscribeProposal returns the MsgProposal subscription for given view(s)
+	SubscribeProposal(sequence, round uint64, higherRounds bool) (Subscription[*MsgProposal], func())
+
+	// SubscribePrepare returns the MsgPrepare subscription for given view(s)
+	SubscribePrepare(sequence, round uint64, higherRounds bool) (Subscription[*MsgPrepare], func())
+
+	// SubscribeCommit returns the MsgCommit subscription for given view(s)
+	SubscribeCommit(sequence, round uint64, higherRounds bool) (Subscription[*MsgCommit], func())
+
+	// SubscribeRoundChange returns the MsgRoundChange subscription for given view(s)
+	SubscribeRoundChange(sequence, round uint64, higherRounds bool) (Subscription[*MsgRoundChange], func())
+}
+
+type (
+	// MsgNotification is received from the subscription to indicate a new message
+	MsgNotification[M IBFTMessage] interface {
+		// Unwrap returns all messages that fit the subscription
+		Unwrap() []M
+	}
+
+	MsgNotificationFn[M IBFTMessage] func() []M
+
+	// Subscription is a channel for IBFT message types
+	Subscription[M IBFTMessage] chan MsgNotification[M]
+)
+
+func (r MsgNotificationFn[M]) Unwrap() []M {
+	return r()
+}
+
+// WrapMessages wraps concrete message types into Message type
+func WrapMessages[M IBFTMessage](messages ...M) []Message {
+	wrapped := make([]Message, 0, len(messages))
+	for _, msg := range messages {
+		wrapped = append(wrapped, Message(msg))
+	}
+
+	return wrapped
+}
+
+// SignMsg returns msg with updated Signature field
+func SignMsg[M IBFTMessage](msg M, signer Signer) M {
```

Comment:
You don't need this insane indirection.

You generate the message signature when you want to send the message, and you have custom handlers for each message type (broadcasts)...why not just define a method on the message type called `SignaturePayload` and pipe it into the `Sign` function, inside those custom message type multicast funcs

Reply:
This function is literally called from those very same handlers where you know the exact message type, and yet there is typecasting here -- useless